### ISSUE TITLE
Fill MarkedSource for functions and parameters. [phab:D1181]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1552]
_Created by **zarko** at 2016-12-09 20:59:08_: 

This diff adds new LLVM dependencies. In third_party/llvm/llvm/build,
you need to run
`cmake --build . -- "-j42" clangFormat clangToolingCore`

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-12-09 21:01:09_:

NB: this function was moved from IndexerASTHooks (but was deleted in an intermediate diff).

**Comment by kythe** [comment]
_**zarko** wrote at 2016-12-12 23:53:35_:

Rebase.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-12-12 23:54:05_:

ping

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 00:10:02_:

Is this only used as an argument to `clang::format::reformat()` and then dropped?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 00:10:02_:

'const bool' should just be 'bool'

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 00:10:02_:

Prefer the weakest operation that will work (`push_back`) barring benchmarks and a corresponding comment.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 00:10:02_:

This lambda exceeds a reasonable length for a lambda; I'd suggest moving this, cancel_count and the `nodes` stack to a separate class.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-12-13 00:51:49_:

  - Address comments.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-12-13 00:52:10_:

It's not clear what conditions lead to it getting set, but I suppose it would be reasonable to treat it as an error condition.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 20:59:21_:

(is fine, but I should have been clearer that renaming is "ignored" or something was sufficient, if not explicitly scoping it as well)

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 20:59:21_:

This doesn't need to be a nested class and makes the enclosing class harder to read.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 20:59:21_:

The `std::pair<>` on the other hand probably does make sense as a trivial nested struct.  `first` and `second` are awful names and should be avoided outside of generic code.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-12-13 22:01:25_:

  - Address comments.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 22:43:22_:

Does `isFileId()` return bool? If so, this check combined with the below is deceptive: just check `!range.getBegin().isFileID() || !range.getEnd().isFileID()`.  Maybe:

```
if (!(range.getBegin().isValid() && range.getBegin().isFileID()) ||
    !(range.getEnd().isValid() && range.getEnd().isFileID())) {
  return {};
}
```

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 22:43:22_:

Is there no `MakeIntrusiveRefCntPtr<T>` ?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 22:43:22_:

`child().empty()`?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 22:43:22_:

ditto

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-12-13 22:43:22_:

range-for won't work here?

**Comment by kythe** [comment]
_**zarko** wrote at 2016-12-14 00:39:55_:

  - Address comments.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-12-14 00:41:00_:

In what way is it deceptive?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-12-14 00:41:00_:

It does not appear so.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-12-14 00:41:00_:

Not directly.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-01-03 22:05:13_:

  - Address comments.
  - Address comments.
  - Address comments.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-01-03 22:05:34_:

Rebase; ping.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-01-03 22:44:20_:

maybe it's just me, but I find switch clearer for this sort of thing:
```
switch(DN.getNameKind()) {
  case clang::DeclarationName::NameKind::ObjCZeroArgSelector:
  case clang::DeclarationName::NameKind::ObjCOneArgSelector:
  case clang::DeclarationName::NameKing::ObjCMultiArgSelector:
    return true;
  default:
    return false;
};
```

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-01-03 22:44:20_:

It's an obfuscated way of requiring that both ranges represent `FileIDs`.  Reading comprehension generally works best when it proceeds linearly rather than jumping around.  This requires reading the second conditional to understand the first really just means that both must be `FileIDs`.  Far less relevant is the redundant call to `isFileID`.  The disparate blocks are logically related.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-01-03 22:44:20_:

The preceding block should be a separate function.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-01-03 22:44:20_:

Given that functions returning function pointers require special support (based on comments), we should likely have a test covering that.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-01-04 00:21:01_:

  - Address comments.
  - Address comments.
  - Address comments.
  - Address comments.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-04 00:21:05_:

This checks that (1) the range is valid (there's a simpler way to do this, which I've now used) and (2) the range begins and ends with the same kind of ID. I don't see what's obfuscated about it, and it's useful to separate in case we start caring about macro text in the future. The subsequent check further excludes macros and tests whether the range starts and ends in the same file.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-04 00:21:05_:

It's used to refer to the file being reformatted. I replaced it with a constant.

**Comment by kythe** [comment]
_**shahms** wrote at 2017-01-04 20:23:58_:

Thanks!

**Comment by kythe** [comment]
_**shahms** accepted this revision at 2017-01-04 20:23:58_

**Comment by kythe** [comment]
_**zarko** wrote at 2017-01-05 22:57:17_:

  - Address comments.
  - Address comments.
  - Address comments.
  - Address comments.
  - Reformatting was far too expensive in practice. There is still
    hope (we might be able to collect all signatures in a buffer and
    do a bulk-reformat before emitting data) but it needs future work.
  - Storing source text for functions was also too expensive, so
    we'll need to pretty-print. This is a minimal pretty-printer
    that only emits FQN(args).
  - We need to be a bit more cautious about what is or is not an
    "implicit" template instantiation.

**Comment by kythe** [comment]
_**zarko** closed via commit 149f1e699353e2645ead372ab0e3a28c28e6e7db at 2017-01-05 23:03:25_
